include/picongpu/particles/atomicPhysics/kernel/FixAtomicState.kernel

```diff
@@ -65,22 +69,33 @@ namespace picongpu::particles::atomicPhysics::kernel
                 [&chargeStateOrgaBox, &atomicStateDataBox](T_Worker const& worker, auto& ion)
                 {
                     using ConfigNumber = typename T_AtomicStateDataDataBox::ConfigNumber;
-
                     uint8_t const boundElectrons = u8(ion[boundElectrons_]);
-                    typename ConfigNumber::DataType const configNumber
-                        = atomicStateDataBox.configNumber(ion[atomicStateCollectionIndex_]);
+                    uint8_t const targetChargeState = ConfigNumber::atomicNumber - boundElectrons;
 
-                    uint8_t const atomicPhysicsBoundElectrons = ConfigNumber::getBoundElectrons(configNumber);
+                    constexpr uint32_t numberAtomicStates
+                        = u32(picongpu::traits::GetNumberAtomicStates<T_IonSpecies>::value);
+                    uint32_t const collectionIndex = ion[atomicStateCollectionIndex_];
 
-                    if(boundElectrons != atomicPhysicsBoundElectrons)
+                    // check for default and out of range values
+                    if(collectionIndex >= numberAtomicStates)
                     {
-                        // set atomic state to ground state
-                        uint8_t const targetChargeState = ConfigNumber::atomicNumber - boundElectrons;
-
                         ion[atomicStateCollectionIndex_] = atomicStateDataBox.findGroundState(
                             chargeStateOrgaBox.numberAtomicStates(targetChargeState),
                             chargeStateOrgaBox.startIndexBlockAtomicStates(targetChargeState));
                     }
+                    else
+                    {
+                        typename ConfigNumber::DataType const configNumber
+                            = atomicStateDataBox.configNumber(collectionIndex);
+                        uint8_t const atomicPhysicsBoundElectrons = ConfigNumber::getBoundElectrons(configNumber);
+
+                        if(boundElectrons != atomicPhysicsBoundElectrons)
+                        {
+                            ion[atomicStateCollectionIndex_] = atomicStateDataBox.findGroundState(
```

Comment:
I will accept this fix but will check to reformulate the code in a separate PR.
There are too many code duplications in.

Reply:
refcator PR #5118

Reply:
I wasn't perfectly happy either, but separating it out caused even more overhead. Maybe you will find a better solution.